From ea40ac1f56c76ac522f582893e8415ac3c32543e Mon Sep 17 00:00:00 2001
From: Marcus Comstedt <marcus@mc.pp.se>
Date: Wed, 16 Jul 2025 14:49:17 +0200
Subject: [PATCH 1/3] cmake: Recognize ppc64

---
 cmake/QtToolchainHelpers.cmake | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/cmake/QtToolchainHelpers.cmake b/cmake/QtToolchainHelpers.cmake
index eaa3a43390..5693c4cc42 100644
--- a/cmake/QtToolchainHelpers.cmake
+++ b/cmake/QtToolchainHelpers.cmake
@@ -55,6 +55,8 @@ function(get_gn_arch result arch)
         set(${result} "mips64el" PARENT_SCOPE)
     elseif(arch STREQUAL "riscv64")
         set(${result} "riscv64" PARENT_SCOPE)
+    elseif(arch STREQUAL "power64")
+        set(${result} "ppc64" PARENT_SCOPE)
     else()
         message(FATAL_ERROR "Unknown architecture: ${arch}")
     endif()
-- 
2.49.1


From 06551090d9f0279188b61cb971d6907e0da63aae Mon Sep 17 00:00:00 2001
From: Marcus Comstedt <marcus@mc.pp.se>
Date: Wed, 6 Jan 2021 13:35:58 +0100
Subject: [PATCH 2/3] Fix compositing of kBGRA_8888_SkColorType on big endian

---
 src/core/compositor/display_software_output_surface.cpp | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/src/core/compositor/display_software_output_surface.cpp b/src/core/compositor/display_software_output_surface.cpp
index 2ad608f26f..e7d6896300 100644
--- a/src/core/compositor/display_software_output_surface.cpp
+++ b/src/core/compositor/display_software_output_surface.cpp
@@ -80,7 +80,12 @@ inline QImage::Format imageFormat(SkColorType colorType)
 {
     switch (colorType) {
     case kBGRA_8888_SkColorType:
+#if Q_BYTE_ORDER == Q_LITTLE_ENDIAN
         return QImage::Format_ARGB32_Premultiplied;
+#else
+        // R/B swap needed...
+        return QImage::Format_RGBA8888_Premultiplied;
+#endif
     case kRGBA_8888_SkColorType:
         return QImage::Format_RGBA8888_Premultiplied;
     default:
@@ -100,6 +105,10 @@ void DisplaySoftwareOutputSurface::Device::swapFrame()
     QImage image(reinterpret_cast<const uchar *>(skPixmap.addr()), viewport_pixel_size_.width(),
                  viewport_pixel_size_.height(), skPixmap.rowBytes(),
                  imageFormat(skPixmap.colorType()));
+#if Q_BYTE_ORDER == Q_BIG_ENDIAN
+	if (skPixmap.colorType() == kBGRA_8888_SkColorType)
+	  image = image.rgbSwapped();
+#endif
     if (m_image.size() == image.size()) {
         QRect damageRect = toQt(damage_rect_);
         QPainter painter(&m_image);
-- 
2.49.1


From d9f14c67be59aba4b805b932218103962576d813 Mon Sep 17 00:00:00 2001
From: Marcus Comstedt <marcus@mc.pp.se>
Date: Mon, 21 Jul 2025 00:30:38 +0200
Subject: [PATCH 3/3] Don't use BGRA on big endian

---
 src/core/compositor/native_skia_output_device_opengl.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/core/compositor/native_skia_output_device_opengl.cpp b/src/core/compositor/native_skia_output_device_opengl.cpp
index 7cfc390d8a..16017c51d3 100644
--- a/src/core/compositor/native_skia_output_device_opengl.cpp
+++ b/src/core/compositor/native_skia_output_device_opengl.cpp
@@ -62,7 +62,7 @@ NativeSkiaOutputDeviceOpenGL::NativeSkiaOutputDeviceOpenGL(
     qCDebug(lcWebEngineCompositor, "Native Skia Output Device: OpenGL");
 
     SkColorType skColorType = kRGBA_8888_SkColorType;
-#if BUILDFLAG(IS_OZONE_X11) && QT_CONFIG(xcb_glx_plugin)
+#if BUILDFLAG(IS_OZONE_X11) && QT_CONFIG(xcb_glx_plugin) && Q_BYTE_ORDER == Q_LITTLE_ENDIAN
     if (OzoneUtilQt::usingGLX() && m_contextState->gr_context_type() == gpu::GrContextType::kGL)
         skColorType = kBGRA_8888_SkColorType;
 #endif
-- 
2.49.1

