From db94f934ed125c4e562d0a6778a652d9c11e6936 Mon Sep 17 00:00:00 2001
From: Marcus Comstedt <marcus@mc.pp.se>
Date: Sun, 14 Feb 2021 19:32:08 +0100
Subject: [PATCH] Remove unsupported -client option to JVM

---
 third_party/closure_compiler/compiler.py | 1 -
 1 file changed, 1 deletion(-)

diff --git a/third_party/closure_compiler/compiler.py b/third_party/closure_compiler/compiler.py
index c6368af733..2616cf7d34 100755
--- a/third_party/closure_compiler/compiler.py
+++ b/third_party/closure_compiler/compiler.py
@@ -24,7 +24,6 @@ class Compiler(object):
       _JAVA_PATH,
       "-jar",
       "-Xms1024m",
-      "-client",
       "-XX:+TieredCompilation",
   ]
 
-- 
2.39.3

